relayer-cli/src/commands/tx/client.rs

```diff
@@ -280,7 +291,20 @@ impl TxUpgradeClientsCmd {
         src_chain: Chain,
     ) -> Result<Vec<IbcEvent>, Error> {
         let client = ForeignClient::restore(client_id, dst_chain, src_chain);
-        client.upgrade().map_err(Error::foreign_client)
+
+        let src_upgrade_height = {
+            let src_application_height = match client.src_chain().query_latest_height() {
+                Ok(height) => height,
+                Err(e) => Output::error(format!("{}", e)).exit(),
+            };
+
+            // When the chain is halted, the application height reports a height
+            // 1 less than the halted height
+            src_application_height.increment()
+        };
+        client
+            .upgrade(src_upgrade_height)
+            .map_err(Error::foreign_client)
```

Comment:
Should we do this only once in `upgrade_clients_for_chain()` or even earlier in `run()`?

Reply:
Do you mean the upgrade, or querying the `src_upgrade_height`? In both cases, I don't see how we can do it earlier. How I understand it:

+ `src_upgrade_height`: must be queried once per "source client", so we can't just query it once for all clients
+ `client.upgrade()`: we need to do this once per "source client" as well, so can't be done earlier.

Is this what you meant?

Reply:
Just `src_upgrade_height`. This is the halt height for the chain that is upgraded, let it be `ibc-0`.  And let's assume we only have one other chain `ibc-1`.
When we issue `hermes upgrade clients ibc-0` we want all affected clients hosted on all of the other chains (on `ibc-1` in our example, clients that verify from `ibc-0`), to be upgraded so they can verify headers from the upgraded `ibc-0` chain.  

> + `src_upgrade_height`: must be queried once per "source client chain", so we can't just query it once for all clients

Not sure what "source client chain" is. Is it the chain for which a client verifies? Is it `ibc-0` in the example? In this case it can only be the one specified in the CLI (`ibc-0`). So we can do it in `TxUpgradeClientsCmd::run()`:
```
        let src_upgrade_height = {
            let src_application_height = match src_chain.query_latest_height() {
                Ok(height) => height,
                Err(e) => Output::error(format!("{}", e)).exit(),
            };

            // When the chain is halted, the application height reports a height
            // 1 less than the halted height
            src_application_height.increment()
        };
```
> + `client.upgrade()`: we need to do this once per "source client chain" as well, so can't be done earlier.

`client.upgrade()` should still be called for all clients on the other chains, `ibc-1` in this example.

Reply:
btw just tried computing `src_upgrade_height` once for all,  [these](https://gist.github.com/ancazamfir/4586f5f8a535b30f2636552bbe969e9d) diffs seem ok

Reply:
Yep you're right, I got confused between source/dest again... Can't wait for `host` & `reference` naming 😅

Reply:
[49df45f](https://github.com/informalsystems/ibc-rs/pull/2304/commits/49df45f7cac1934b116349e3006036df4f1aef51)